2/contrib/jenkins/install-plugins.sh

```diff
@@ -233,22 +235,55 @@ function resolveDependencies() {
+        #
+        local versionFromPluginParam
         if [[ $d == *"resolution:=optional"* ]]; then
-            echo "Skipping optional dependency $plugin"
-        else
-            local pluginInstalled
-            if pluginInstalled="$(echo "${bundledPlugins}" | grep "^${plugin}:")"; then
-                pluginInstalled="${pluginInstalled//[$'\r']}"
-                local versionInstalled; versionInstalled=$(versionFromPlugin "${pluginInstalled}")
-                local minVersion; minVersion=$(versionFromPlugin "${d}")
-                if versionLT "${versionInstalled}" "${minVersion}"; then
-                    echo "Upgrading bundled dependency $d ($minVersion > $versionInstalled)"
-                    download "$plugin" &
-                else
-                    echo "Skipping already bundled dependency $d ($minVersion <= $versionInstalled)"
-                fi
+            #echo "Skipping optional dependency $plugin"
+            versionFromPluginParam="$(cut -d';' -f1 - <<< "$d")"
+	else
+            versionFromPluginParam=$d
+	fi
+        local pluginInstalled
+        local minVersion; minVersion=$(versionFromPlugin "${versionFromPluginParam}")
+
+	set +o pipefail
+	local filename; filename=$(getArchiveFilename "$plugin")
+	local previouslyDownloadedVersion; previouslyDownloadedVersion=$(get_plugin_version $filename)
+	set -o pipefail
+	
+        # ${bundledPlugins} checkds for plugins bundled in the jenkins.war file; per 
+        # https://wiki.jenkins-ci.org/display/JENKINS/Bundling+plugins+with+Jenkins this is getting
+        # phased out, but we are keeping this check in for now while that transition bakes a bit more	
+        if pluginInstalled="$(echo "${bundledPlugins}" | grep "^${plugin}:")"; then
+            pluginInstalled="${pluginInstalled//[$'\r']}"
+            # get the version of the plugin bundled
+            local versionInstalled; versionInstalled=$(versionFromPlugin "${pluginInstalled}")
+            # if the bundled plugins is older than the minimum version needed for the dependency,
+            # download the dependence; passing "true" is needed for "download" to replace the existing dependency
+            if versionLT "${versionInstalled}" "${minVersion}"; then
+                echo "Upgrading bundled dependency $d ($minVersion > $versionInstalled)"
+                download "$plugin" "$minVersion" "true"
             else
-                download "$plugin" &
+                echo "Skipping already bundled dependency $d ($minVersion <= $versionInstalled)"
+            fi
+            # bypass further processing if a bundled plugin
+            continue
+        fi
+
+        # if the dependency plugin has yet to be downloaded (hence the var is not set) download
+        if [[ -z "${previouslyDownloadedVersion:-}" ]]; then
+            echo "Downloading previously uninstalled dependency plugin $plugin version $minVersion"
```

Comment:
does "previously uninstalled" mean "previously not installed"?  i keep reading it as "downloading a plugin that was installed at some point previously, then uninstalled", which made me very confused.  If i'm right in my new understanding, can we rephrase? :)